graph.py

```diff
@@ -172,108 +243,128 @@ class ActionList:
+        ctx, model_action.id, model_group_action.group_id, model_action.name, model_action.display_name,
         model_action.description, model_action.hidden, model_group_action.dynamic
     )
     key = (action.group_id, action.action_id)
+    ctx.actions[key] = action
     for affect in models.group_action_affects[key]:
         if affect.task == Task.START:
-            action.start_affects.extend(_build_affects(affect))
+            action.start_affects.extend(_build_affects(affect, ctx))
         if affect.task == Task.COMPLETE:
-            action.complete_affects.extend(_build_affects(affect))
+            action.complete_affects.extend(_build_affects(affect, ctx))
 
     for model_action_email in models.action_emails[action.action_id]:
+        email = _build_email(models, model_action_email, action, ctx)
         if model_action_email.task == Task.START:
-            action.start_emails.append(
-                Email(
-                    action.group_id, action.action_id, models.emails[model_action_email.email_id].name,
-                    model_action_email.task
-                )
-            )
+            action.start_emails.append(email)
         if model_action_email.task == Task.COMPLETE:
-            action.complete_emails.append(
-                Email(
-                    action.group_id, action.action_id, models.emails[model_action_email.email_id].name,
-                    model_action_email.task
-                )
-            )
+            action.complete_emails.append(email)
+
+    _add_partner_restrictions(models, action, models.group_action_partner_restrictions[key])
+
     return action
 
+def _add_partner_restrictions(models, restricted, restrictions):
+    for restriction in restrictions:
+        partner =_build_partner(models, models.partners[restriction.partner_id])
+        if restriction.include:
+            restricted.required.append(partner)
+        else:
+            restricted.excluded.append(partner)
 
-def _build_group(models, model_alist_group):
+def _build_group(models, model_alist_group, ctx):
     model_group = models.groups[model_alist_group.group_id]
     group = Group(model_group.id, model_group.name, model_alist_group.optional)
+    ctx.groups[group.id] = group
     for model_group_action in models.group_actions[group.id]:
-        group.actions.append(_build_action(models, model_group_action))
+        group.actions.append(_build_action(models, model_group_action, ctx))
     for model_trigger in models.triggers[group.id]:
-        group.triggers.extend(_build_triggers(models, model_trigger))
+        group.triggers.extend(_build_triggers(models, model_trigger, ctx))
+    _add_partner_restrictions(models, group, models.group_partner_restrictions[group.id])
     return group
 
 
 def build_action_list(models, action_list_id):
     model_alist = models.action_lists[action_list_id]
 
     # Build the structure of all the groups, actions, affects, triggers, and emails
+    ctx = Context()
```

Comment:
This context class is a bit cryptic, and is getting modified down the call tree.  Is there a way to describe it more explicitly, and maybe build it up separately from it's usage?